crates/devolutions-pedm/src/account.rs

```diff
@@ -0,0 +1,1006 @@
+            return Err(ParseSidError::InvalidFormat);
+        };
+        let domain_id = DomainId::from_str(domain_id)?;
+        let rid = rid.parse().map_err(|e| ParseSidError::ParseInt {
+            source: e,
+            field: "relative_id",
+            value: rid.to_owned(),
+        })?;
+        Ok(Self {
+            domain_id,
+            relative_id: rid,
+        })
+    }
+}
+
+impl fmt::Display for Sid {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "S-1-5-{}-{}", self.domain_id, self.relative_id)
+    }
+}
+
+/// Error type for parsing a SID or domain ID.
+#[derive(Debug)]
+pub enum ParseSidError {
+    InvalidFormat,
+    MissingPrefix,
+    ParseInt {
+        source: ParseIntError,
+        field: &'static str,
+        value: String,
+    },
+    DomainId(ParseDomainIdError),
+}
+
+impl error::Error for ParseSidError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            Self::InvalidFormat | Self::MissingPrefix => None,
+            Self::ParseInt { source, .. } => Some(source),
+            Self::DomainId(e) => Some(e),
+        }
+    }
+}
+
+impl fmt::Display for ParseSidError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::InvalidFormat => write!(f, "invalid format'"),
+            Self::MissingPrefix => write!(f, "missing prefix"),
+            Self::ParseInt { source, field, value } => {
+                write!(f, "failed to parse field {} with value {}: {}", field, value, source)
+            }
+            Self::DomainId(e) => e.fmt(f),
+        }
+    }
+}
+
+impl From<ParseDomainIdError> for ParseSidError {
+    fn from(e: ParseDomainIdError) -> Self {
+        Self::DomainId(e)
+    }
+}
+
+/// A user account.
+///
+/// This contains the information that is retrieved from WinAPI.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct Account {
+    pub(crate) name: String,
+    pub(crate) sid: Sid,
+}
+
+/// A user account with a unique ID.
+///
+/// This is the type to use when referring to accounts.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct AccountWithId {
+    /// The numeric ID for the account ID that was generated by the database.
+    pub(crate) id: i16,
+    pub(crate) name: String,
+    /// The numeric ID for the domain that was generated by the database.
+    pub(crate) internal_domain_id: i16,
+    pub(crate) sid: Sid,
+}
+
+impl PartialEq<Account> for AccountWithId {
+    fn eq(&self, other: &Account) -> bool {
+        self.name == other.name && self.sid == other.sid
+    }
+}
+
+/// Get the list of usernames and corresponding SIDs.
+///
+/// `LookupAccountNameW` must be called to enable `ConvertSidToStringSidW` to work.
+#[cfg(target_os = "windows")]
+#[allow(clippy::multiple_unsafe_ops_per_block)]
+#[allow(clippy::cast_possible_truncation)]
+pub(crate) unsafe fn list_accounts() -> Result<Vec<Account>, ListAccountsError> {
```

Comment:
question: Why this function is marked as unsafe? What are the safety preconditions to uphold in order to call it? It seems to me that it is wrapping unsafe operations in a safe way (unless there are other "ambient" preconditions?).